onnxruntime/test/providers/cpu/tensor/gather_nd_op_test.cc

```diff
@@ -160,9 +168,57 @@ TEST(GatherNDOpTest, GatherND_slice_float_batch_dims_2) {
   test.Run();
 }
 
+TEST(GatherNDOpTest, GatherND_slice_float_batch_dims_one_3) {
+  OpTester test("GatherND", 12, kOnnxDomain);
+  test.AddAttribute<int64_t>("batch_dims", 1);
+  test.AddInput<float>("data", {2, 2, 2}, ValueRange(8, 0.0f, 0.1f));
+  test.AddInput<int64_t>("indices", {2, 1, 2}, {1LL, 0LL, 0LL, 1LL});
+  test.AddOutput<float>("output", {2, 1}, {0.2f, 0.5f});
+  test.Run();
+}
+
+TEST(GatherNDOpTest, GatherND_slice_float_batch_dims_two) {
+  OpTester test("GatherND", 12, kOnnxDomain);
+  test.AddAttribute<int64_t>("batch_dims", 2);
+  test.AddInput<float>("data", {2, 1, 3, 5}, ValueRange(30, 0.0f, 0.1f));
+  test.AddInput<int64_t>("indices", {2, 1, 3, 2},
+                         {0LL, 0LL,
+                          0LL, 1LL,
+                          1LL, 0LL,
+                          1LL, 1LL,
+                          0LL, 4LL,
+                          2LL, 4LL});
+  test.AddOutput<float>("output", {2, 1, 3}, {0.0f, 0.1f, 0.5f, 2.1f, 1.9f, 2.9f});
+  test.Run();
+}
+
+TEST(GatherNDOpTest, GatherND_negative_slice_float_batch_dims_one) {
+  OpTester test("GatherND", 12, kOnnxDomain);
+  test.AddAttribute<int64_t>("batch_dims", 1);
+  test.AddInput<float>("data", {2, 3, 4}, ValueRange(24, 1.0f));
+  test.AddInput<int64_t>("indices", {2, 2, 2}, {0LL, -3LL, -1LL, 2LL, -1LL, 0LL, 0LL, -2LL});
+  test.AddOutput<float>("output", {2, 2}, {2.0, 11.0, 21.0, 15.0});
+  test.Run();
+}
+
+TEST(GatherNDOpTest, GatherND_negative_slice_float_batch_dims_two) {
+  OpTester test("GatherND", 12, kOnnxDomain);
+  test.AddAttribute<int64_t>("batch_dims", 2);
+  test.AddInput<float>("data", {2, 1, 3, 5}, ValueRange(30, 0.0f, 0.1f));
+  test.AddInput<int64_t>("indices", {2, 1, 3, 2},
+                         {0LL, -5LL,
+                          -3LL, 1LL,
+                          -2LL, 0LL,
+                          -2LL, -4LL,
+                          0LL, -1LL,
+                          2LL, -1LL});
+  test.AddOutput<float>("output", {2, 1, 3}, {0.0f, 0.1f, 0.5f, 2.1f, 1.9f, 2.9f});
+  test.Run();
+}
+
 #ifdef USE_CUDA
 #if __CUDA_ARCH__ >= 600
```

Comment:
looks like `__CUDA_ARCH__` won't be defined in host code: https://stackoverflow.com/questions/16072740/the-behavior-of-cuda-arch-macro
perhaps this is not the right check

Reply:
Yeah I noticed there was some conversations about this earlier. Do you know what part of GatherND cuda kernels require __CUDA_ARCH__ >= 600? I was hitting "invalid device" error for GPU ci test for double GatherND, and fixed it adding "if (!HasCudaEnvironment(600 /*min_cuda_architecture*/)) return;" for that test. Did not know why.

Reply:
It seems __CUDA_ARCH__ is only defined in the .cu file ... 
I wonder what would it return in the CPU only build? Will these tests be skipped ? 
if (!HasCudaEnvironment(600 /min_cuda_architecture/))

Reply:
Yes, CUDA_ARCH is only defined for .cu files.

Reply:
To be more clear, according to CUDA programming guide, the host code (the non-GPU code) must not depend on __CUDA_ARCH__. My interpretation is that __CUDA_ARCH__ can only be used in GPU device code. But, I found CUDA_ARCH is defined for .cu files. So, it seems OK to use CUDA_ARCH to differentiate different code path in .cu files for now.

The below is excerpted from CUDA doc for your reference.

The architecture identification macro __CUDA_ARCH__ is assigned a three-digit value string xy0 (ending in a literal 0) during each nvcc compilation stage 1 that compiles for compute_xy. This macro can be used in the implementation of GPU functions for determining the virtual architecture for which it is currently being compiled. The host code (the non-GPU code) must not depend on it.

Reply:
@SherlockNoMad for those test protected by following code, CPU only build will skip. 

if (!HasCudaEnvironment(600 /*min_cuda_architecture*/)) return;

Reply:
Thanks @weixingzhang for the explanation. For this specific place  I'll use HasCudaEnvironment to replace CUDA_ARCH.

Reply:
I think our CI machine doesn't have arch>600. 
This is means that these tests will not be ran in both CPU and GPU builds. 

 Shall we make HasCudaEnvironment() -> IsCPUBuildOrHasCudaEnvironment() ?

Reply:
@SherlockNoMad Got your point. maybe we can add a new function for that, then CPU build will not skip those tests any more.

Reply:
while, any good way to check CPU only build besides this way? 

inline bool IsCPUBuild() {
#ifdef USE_CUDA
  return false;
#endif
  return true;
}